### DISCUSSION TITLE
Separating outputs for multi-label segmentation

### DISCUSSION THREAD

**Discussion by ghost** [Project-MONAI/MONAI#2570] [Q&A]
Hi,
I have labels for two features I am segmenting and for later processing it would be useful to export the model segmentations separately. If I'm missing a tutorial/example for how to do this please let me know but here is what I've tried so far:

The shape of the whole segmentation is:
>torch.Size([1, 2, 64, 64, 64])

Therefore I have tried iterating over the second element to produce separated outputs with:

```
model.load_state_dict(torch.load(trained_model))
model.eval()
    with torch.no_grad():
        for image in range(len(case_name)):
            print(f'Processing {case_name[image]}')

            val_input = val_ds[image]["image"].unsqueeze(0).to(device)
            post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True, logit_thresh=0.7)]     )  
    
            val_output = model(val_input)
            val_output_bin = post_trans(val_output)

             # separate outputs for the two labels for features LP and TY
             feature = ['LP', 'TY']
             for anatomy in range(2):
                  print(val_output_bin_largest.shape)
                  saver = NiftiSaver(output_dir=outputdata_dir+ f'/{case_name[image]}', output_postfix=f'{feature[anatomy]}_seg', dtype=(['numpy.float16']))
                  saver.save_batch(val_output_bin[:, anatomy, :, :, :], meta_data = {'filename_or_obj': case_name})
```

This does output the labels separately, but in a different orientation to the un-separated output `saver.save_batch(val_output_bin)` and this orientation is not ideal for future processing steps. I know I could do some transforms but am I missing something? I had hoped the separate segmentations would simply output in the same orientation as separate files.

Edit:
I realised it would be useful to share the different output shapes.
Combined output:(64, 64, 64, 2)
Separated outputs: (64, 64, 1, 64)

Any assistance would be greatly appreciated as always! Thank you in advance

**Comment by Nic-Ma** [comment]
Hi @jnc13 ,

Thanks for your interest here.
I didn't quite understand your question, but I think maybe you want to add `argmax` to the segmentation, you can set it by `AsDiscrete` transform.
If you want to split channels, you can use `SplitChannel` transform.
@rijobro please feel free to correct me if I misunderstand anything here.

Thanks.

**Reply by ghost** [reply]
Thanks as always @Nic-Ma!

I've now added a SplitChannel transform, and have adjusted the `channel_dim` but am struggling to get it to do exactly what I'm after.

Basically I'm segmenting two features and want to split the outputs to two files. It sounds like SplitChannel could help but the output is not what I expect.
The initial segmentation has a shape of torch.Size([1, 2, 64, 64, 64]) with the 2nd dimension (or 1st with 0 index) being the one with the two feature channels. With no further processing, when I save this data I get an output image of shape (64, 64, 64, 2). All I want to be able to do is end up with two images, one for each of the two features eg of shape 64,64,64,1 and 64,64,64,1

Going back to the shape of the segmentation ([1, 2, 64, 64, 64]), I think I need to set `channel_dim=(1)` to split in the "2" channel. Doing so I get three outputs per image: 0,1,2. None of the three are what I was hoping for and don't seem to split the shape as I hoped. Please could you advise what I'm doing wrong?

Edit: When I set  `channel_dim=(1)`, the first two channels are clearly split in the wrong axis and the final of the three saved files is in fact the same as the non-transformed saved image (so still split across two channels).

Thank you as always!

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

May I know your MONAI version and could you please help share your latest test program?

Thanks.

**Reply by ghost** [reply]
Hi @Nic-Ma ,

I'm using version0.5.3 (transitioning to 0.6!). Here is my test program I use for segmentation once I have the trained model. The bit of interest is right at the bottom:

```py
class ConvertToMultiChannelBasedOnClassesd(MapTransform):
    """
    Convert labels:
    label 1 is feat1
    label 2 is feat2
    """
    def __call__(self, data):
        d = dict(data)
        for key in self.keys:
            result = []
            # label 1 is feat1
            result.append(d[key] == 1)
            # label 2 is feat2
            result.append(d[key] == 2)
            d[key] = np.stack(result, axis=0).astype(np.float32)
        return d  

def feat_auto_segmentation(input_images, trained_model, voxel_size, image_type, input_labels = 'NA'):
    """
    When given an input CT .nii image, uses our optimised segmentation model
    to output both segmented .nii images and figures showing the comparison of input scan and segmentation

    Parameters
    ----------
    input_images : CT Nifti image .nii
    trained_model : PyTorch model .pth
    voxel_size :
    image_type : validation or test? Dictates if the script will use ground-truth labels
    input_labels (optional) : If for validation images, provide location of ground-truth labels

    Returns
    -------
    None.

    """  
    hyperparameter_config = dict (
##Some values I use for settings
    )

    config=hyperparameter_config
    
    data_dir = input_images
    outputdata_dir = input_images

    fresh_images = sorted(glob.glob(os.path.join(data_dir, "*.nii.gz")))
    #Exclude segmented images if using this function on second iteration
    included_cases = []
    for case in fresh_images:
        if (case.endswith("seg.nii.gz") == False):
            included_cases.append(case)
        fresh_images = included_cases
    
    #Make dictionary of case names
    case_name = []
    # return all files as a list
    for file in os.listdir(input_images):
     # check the files which are end with specific extention
             if file.endswith(".nii.gz") and (file.endswith("seg.nii.gz") == False) :
             # print path name of selected files to txt file
                 print('Case: ', file[:-7])
                 case_name.append(file[:-7])
                 patient_dict = {
                     'name': file[:-7],
                 }
                 f = open(input_images+f"/{file[:-7]}.txt", "w")
                 f.write(str(f'name {file[:-7]} \n'))
                 f.close()
                 
    print('List of cases: ', case_name)
    print(f"Number of datasets received: {len(fresh_images)}")
    
    #Can either segment "test" images for which there is no manual ground-truth label
    #Or "validation" images for which we have the segmentation label available
     
    if image_type == 'Test' :
        data_dicts = [{"image": image_name, "label": label_name} for image_name, label_name in zip(fresh_images, fresh_images)]
    
    elif image_type == 'Validation' :
        val_labels = sorted(glob.glob(os.path.join(input_labels, "*.nii.gz")))
        data_dicts = [
            {"image": image_name, "label": label_name}
            for image_name, label_name in zip(fresh_images, val_labels)
            ]
    else:
        print('Provide a suitable image type arguement: "Test" or "Validation"')

    val_files = data_dicts[:]

    #Transformations must match those used during training
    val_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        AddChanneld(keys=["image"]),
        ConvertToMultiChannelBasedOnClassesd(keys="label"),
        Spacingd(
            keys=["image", "label"],
            pixdim=(config['VoxelSize'], config['VoxelSize'], config['VoxelSize']),
            mode=("bilinear", "nearest")
        ),
        Orientationd(keys=["image", "label"], axcodes="PLS"), ##Use LPS finally
        CenterSpatialCropd(keys=["image", "label"],roi_size=[config['RandCropX'], config['RandCropY'], config['RandCropZ']]),
        NormalizeIntensityd(keys="image", nonzero=True, channel_wise=True),
        ToTensord(keys=["image", "label"]),
    ]
    )

    val_ds = CacheDataset(data=val_files, transform=val_transforms, cache_rate=1.0, num_workers=0)
    val_loader = DataLoader(val_ds, batch_size=1, num_workers=0)

    set_determinism(seed=0)
    
    
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    model = UNet(
        dimensions=3,
        in_channels=1,
        out_channels=2,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
    ).to(device)
    
    loss_function = DiceLoss(to_onehot_y=False, sigmoid=True, squared_pred=True)

    optimizer = torch.optim.Adam(model.parameters(), 1e-4, weight_decay=1e-5, amsgrad=True)

    post_pred = AsDiscrete(argmax=True, to_onehot=True, n_classes=2)
    post_label = AsDiscrete(to_onehot=True, n_classes=2)

    
    #Now actually carry out the segmentation and image saving
    
    model.load_state_dict(torch.load(trained_model))
    model.eval()
    

    with torch.no_grad():
        for image in range(len(case_name)):
            print(f'Processing {case_name[image]}')
            
            val_input = val_ds[image]["image"].unsqueeze(0).to(device)
            post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True, logit_thresh=0.99)] #logit_thresh= 0.9 seems good currently
                                     )  
    
            val_output = model(val_input)
            val_output_bin = post_trans(val_output)   
            val_output_bin_largest = KeepLargestConnectedComponent(applied_labels=[0,1])(val_output_bin)

                        
            # Overlay of largest component- model labels and input image
            # Visualise overlay of two model segmentation channels on input image
            plt.figure("Model Segmentations Largest Overlay", (4,4))
            plt.title(f"gt label largest overlay {case_name[image]}")
            # Scan image
            plt.imshow(val_ds[image]["image"][0, :, :, int(slices_aug[image])].detach().cpu(), cmap="gray") # image
            # feat1
            plt.imshow(val_output_bin_largest[0, 0, :, :, int(slices_aug[image])], alpha=val_output_bin_largest[0, 0, :, :, int(slices_aug[image])], cmap="Blues")
            # feat2
            plt.imshow(val_output_bin_largest[0, 1, :, :, int(slices_aug[image])], alpha=val_output_bin_largest[0, 1, :, :, int(slices_aug[image])], cmap="plasma")
            plt.savefig(input_images+f'/{case_name[image]}_model_overlay_largest.png')
            plt.show()


            print('Now output segmentations')
            feature = ['feat1', 'feat2']
            post_trans_split = Compose(SplitChannel(
                channel_dim=1    )
                                     )  
            print('Split complete')
            val_output_bin_largest_split = post_trans_split(val_output_bin_largest)
            saver = NiftiSaver(output_dir=outputdata_dir+ f'/{case_name[image]}', output_postfix=f'largest_split_{feature[anatomy]}_seg', dtype=(['numpy.float16']))
            saver.save_batch(val_output_bin_largest_split,
                             )

    print("All segmentations complete.")
```

Thanks as always!

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

I think the root cause is that: the output of `SplitChannel` transform is a list of `batch-first` tensors, every item is a class.
So when calling `NiftiSaver`, you need to `save_batch` for every class separately I think:
```py
val_output_bin_largest_split = SplitChannel(channel_dim=1)(val_output_bin_largest)

for i, c in enumerate(val_output_bin_largest_split):
    saver = NiftiSaver(output_dir=XXX, output_postfix=f'class_{i}')
    saver.save_batch(c)
```

Thanks.

**Reply by ghost** [reply]
Yep, that solved it. Now working exactly as I hoped it would.

Thank you so so much @Nic-Ma , not just for answering this question but your continual help and support. It makes MONAI so much more approachable for those of us who want to apply the technology to our medical research but don't have a background in coding. So thank you :)

**Reply by Nic-Ma** [reply]
No problem, glad to see your good progress.

Thanks.

**Reply by oscarramirezs** [reply]
hi Nic 
Im trying to fo the same 

      vali_transforms = Compose(
          [
              LoadImaged(keys=["image"]),
              EnsureChannelFirstd(keys=["image"]),
              Orientationd(keys=["image"], axcodes="RAS"),      
              EnsureTyped(keys=["image"]),
          ]
      )


    with torch.no_grad():
            for val_data in vali_loader:
                
                val_inputs = val_data["image"].to(device)
                
                roi_size = (512, 512, 256)
                
                sw_batch_size = 1
                
                val_outputs = sliding_window_inference(
                    val_inputs, roi_size, sw_batch_size, model_scapula_humerus)
                
                val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
                
                val_output_bin_largest_split = SplitChannel(channel_dim=1)(val_outputs)
                
                for i, c in enumerate(val_output_bin_largest_split):
                    saver = NiftiSaver(output_dir="./segmentation_output", output_postfix=f'class_{i}')
                    saver.save_batch(c)

                     
    
but I got the following erreur 

      AttributeError                            Traceback (most recent call last)
      <ipython-input-76-acde0dae54fd> in <module>
            7                 val_inputs, roi_size, sw_batch_size, model_scapula_humerus)
            8             val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
      ----> 9             val_output_bin_largest_split = SplitChannel(channel_dim=1)(val_outputs)
           10             for i, c in enumerate(val_output_bin_largest_split):
           11                 saver = NiftiSaver(output_dir="./segmentation_output", output_postfix=f'class_{i}')
      
      C:\ProgramData\Anaconda3\lib\site-packages\monai\transforms\utility\array.py in __call__(self, img)
          295 
          296     def __call__(self, img: NdarrayOrTensor) -> List[NdarrayOrTensor]:
      --> 297         num_classes = img.shape[self.channel_dim]
          298         if num_classes <= 1:
          299             raise RuntimeError("input image does not contain multiple channels.")
      
      AttributeError: 'list' object has no attribute 'shape'

thank you for your help 

Oscar

**Reply by KumoLiu** [reply]
Hi @oscarramirezs, `decollate_batch` is to de-collate a batch of data, so put it out of loop like this:
```
for val_data in vali_loader:
            
    val_inputs = val_data["image"].to(device)
            
    roi_size = (512, 512, 256)
            
    sw_batch_size = 1
            
    val_outputs = sliding_window_inference(
        val_inputs, roi_size, sw_batch_size, model_scapula_humerus)
            
val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
            
val_output_bin_largest_split = [SplitChannel(channel_dim=1)(i) for i in val_outputs]
```

**Reply by oscarramirezs** [reply]
Hi @KumoLiu thank you for your help 

    with torch.no_grad():
            for val_data in vali_loader:
                
                val_inputs = val_data["image"].to(device)
                
                roi_size = (512, 512, 224)
                
                sw_batch_size = 1
                
                val_outputs = sliding_window_inference(
                    val_inputs, roi_size, sw_batch_size, model_scapula_humerus)
                
            val_outputs = [post_trans(i) for i in decollate_batch(val_outputs)]
            
                
            val_output_bin_largest_split = [SplitChannel(channel_dim=1)(i) for i in val_outputs]
     
            for i, c in enumerate(val_output_bin_largest_split):
                saver = NiftiSaver(output_dir="./segmentation_output", output_postfix=f'class_{i}')
                saver.save_batch(c)


I don't get a batch but a file for each image.
you should get only two files. and thank you again 

    file written: segmentation_output\0\0_class_0.nii.gz.
    file written: segmentation_output\1\1_class_0.nii.gz.
    file written: segmentation_output\2\2_class_0.nii.gz.
    file written: segmentation_output\3\3_class_0.nii.gz.
    file written: segmentation_output\4\4_class_0.nii.gz.
    file written: segmentation_output\5\5_class_0.nii.gz.
    file written: segmentation_output\6\6_class_0.nii.gz.
    file written: segmentation_output\7\7_class_0.nii.gz.
    file written: segmentation_output\8\8_class_0.nii.gz.
    file written: segmentation_output\9\9_class_0.nii.gz.
    file written: segmentation_output\10\10_class_0.nii.gz.
    file written: segmentation_output\11\11_class_0.nii.gz.
    file written: segmentation_output\12\12_class_0.nii.gz.
    file written: segmentation_output\13\13_class_0.nii.gz.
    file written: segmentation_output\14\14_class_0.nii.gz.
    file written: segmentation_output\15\15_class_0.nii.gz.
    file written: segmentation_output\16\16_class_0.nii.gz.
    file written: segmentation_output\17\17_class_0.nii.gz.
    file written: segmentation_output\18\18_class_0.nii.gz.
    file written: segmentation_output\19\19_class_0.nii.gz.
    file written: segmentation_output\20\20_class_0.nii.gz.
    file written: segmentation_output\21\21_class_0.nii.gz.
    file written: segmentation_output\22\22_class_0.nii.gz.
    file written: segmentation_output\23\23_class_0.nii.gz.
    file written: segmentation_output\24\24_class_0.nii.gz.
    file written: segmentation_output\25\25_class_0.nii.gz.
    file written: segmentation_output\26\26_class_0.nii.gz.
    file written: segmentation_output\27\27_class_0.nii.gz.
    file written: segmentation_output\28\28_class_0.nii.gz.
    file written: segmentation_output\29\29_class_0.nii.gz.
    file written: segmentation_output\30\30_class_0.nii.gz.
    file written: segmentation_output\31\31_class_0.nii.gz.
